### DISCUSSION TITLE
[jsonnet] How to add child routes in the AlertManager configuration?

### DISCUSSION THREAD

**Discussion by BeyondEvil** [prometheus-operator/prometheus-operator#4983] [Q&A] [answered]
I'm using the latest (0.58) generated prometheus-operator jsonnet library to configure AlertManager.

And I'm trying to add a "blackhole"-route for watchdog, but to make it work, I end up having to do `blackhole_route.spec.route` which just feels wrong.

What am I not understanding?

```jsonnet
      local alertmanagerConfig = prometheus.monitoring.v1alpha1.alertmanagerConfig;
      local receivers = alertmanagerConfig.spec.receivers;
      local route = alertmanagerConfig.spec.route;

      local blackhole_route =
        route.withReceiver('blackhole')
        + route.withMatchers(
          route.matchers.withName('alertname')
          + route.matchers.withValue('Watchdog')
          + route.matchers.withMatchType('=')
        );

      local slack_receiver =
        local slackConfigs = receivers.slackConfigs;
        receivers.withName('default')
        + receivers.withSlackConfigs(
          slackConfigs.apiURL.withName('alertmanager')
          + slackConfigs.apiURL.withKey('slack_api_url')
          + slackConfigs.withSendResolved(true)
          + slackConfigs.withTitleLink(<url>)
          + slackConfigs.withTitle(<template>)
          + slackConfigs.withText(<template>);

      alertmanagerConfig.new('main-config')
      + alertmanagerConfig.metadata.withLabels({ alertmanagerConfig: 'main' })
      + alertmanagerConfig.spec.withReceiversMixin([
        receivers.withName('blackhole'),
        slack_receiver,
      ])
      + route.withReceiver('default')
      + route.withGroupByMixin(['alertname'])
      + route.withRoutes([blackhole_route.spec.route]),
```

[CRD](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/crds/crd-alertmanagerconfigs.yaml)

[jsonnet](https://github.com/jsonnet-libs/prometheus-operator-libsonnet/blob/main/0.58/_gen/monitoring/v1alpha1/alertmanagerConfig.libsonnet)

**Comment by Duologic** [comment, marked as answer]
It depends on your intention, the generated jsonnet for `route` is intended to configure `alertmanagerConfig.spec.route`, not necessarily the subroutes. With outher words `alertmanagerConfig.spec.route` is not a first-class citizen in that library. 

It will nest the route in `<object>.spec.route` where object is intended to be of type `alertmanagerConfig` object, not a `route`, this type change makes it feel wrong.

However as the generated json from this is quite consistent, I don't consider it wrong, a bit odd at best. Perhaps you could do this so `blackhole_route` actually contains an object of type `route`:

```
local blackhole_route = (
  route.withReceiver('blackhole')
  + route.withMatchers(
    route.matchers.withName('alertname')
    + route.matchers.withValue('Watchdog')
    + route.matchers.withMatchType('=')
  )
).spec.route;
```

**Reply by BeyondEvil** [reply]
Appreciate the response @Duologic 🙏 

Would it make sense to add a customization to the generated library, so that you can add objects of type `route`?

Subroutes/childroutes are part of the spec after all.

**Reply by Duologic** [reply]
Sure, I have no objections.

**Reply by BeyondEvil** [reply]
Any idea about how to go about that in a more dynamic way?

Basically, all the first-level functions from `alertmanagerConfig.spec.route` needs to be duplicated without `spec.route`.

`alertmanagerConfig.spec.route.matchers` is ok tho.

**Reply by Duologic** [reply]
You could do something like this:

```jsonnet
local route = std.foldl(
  function(acc, k)
    local item = alertmanagerConfig.spec.route[k];
    acc +
    if std.isFunction(item)
    then { [k](arg): item(arg).spec.route }
    else { [k]: item },
  std.objectFields(alertmanagerConfig.spec.route),
  {}
);
```

But that really looks ugly and far less elegant than `(...).spec.route`.
